constraints/miden-vm/decoder.air

```diff
@@ -0,0 +1,534 @@
+    let u_respan = f_respan(op_bits, extra) * h_abp
+
+    # When END operation is executed, the hash result is copied into registers 
+    # hasher[0], ..., hasher[3].
+    let u_end = f_end(op_bits, extra) * h_res
+
+    # Enforce the block hash computation constraint.
+    enf p[0]' * (u_ctrli + u_span + u_respan + u_end + 1 - (f_ctrli(op_bits, extra) + 
+        f_span(op_bits) + f_respan(op_bits, extra) + f_end(op_bits, extra))) = p[0]
+
+
+# Enforces constraint for updating the block stack table.
+#
+# Constraint degree: 8
+ev block_stack_table_constraints([addr, op_bits[7], hasher[8], s0], [extra, p[4]]):
+    # When JOIN operation is executed, row (addr', addr, 0) is added to the block stack table.
+    let v_join = f_join(op_bits) * (alpha[0] + alpha[1] * addr' + alpha[2] * addr)
+
+    # When SPLIT operation is executed, row (addr', addr, 0) added to the block stack table.
+    let v_split = f_split(op_bits) * (alpha[0] + alpha[1] * addr' + alpha[2] * addr)
+
+    # When LOOP operation is executed, row (addr', addr, 1) is added to the block stack table if 
+    # the value at the top of the operand stack is 1, and row (addr', addr, 0) is added to the 
+    # block stack table if the value at the top of the operand stack is 0.
+    let v_loop = f_loop(op_bits) * (alpha[0] + alpha[1] * addr' + alpha[2] * addr + alpha[3] * s0)
+
+    # When SPAN operation is executed, row (addr', addr, 0) is added to the block stack table.
+    let v_span = f_span(op_bits) * (alpha[0] + alpha[1] * addr' + alpha[2] * addr)
+
+    # When RESPAN operation is executed, row (addr, hasher[1]', 0) is removed from the block stack 
+    # table, and row (addr', hasher[1]', 0) is added to the table. The prover sets the value of 
+    # register hasher[1] at the next row to the ID of the parent block:
+    let u_respan = f_respan(op_bits, extra) * (alpha[0] + alpha[1] * addr + alpha[2] * hasher[1]')
+    let v_respan = f_respan(op_bits, extra) * (alpha[0] + alpha[1] * addr' + alpha[2] * hasher[1]')
+
+    # When END operation is executed, row (addr, addr', hasher[5]) is removed from the block span 
+    # table. Register hasher[5] contains the is_loop flag.
+    let u_end = f_end(op_bits, extra) * 
+        (alpha[0] + alpha[1] * addr + alpha[2] * addr' + alpha[3] * hasher[5])
+
+    # Enforce the block stack table constraint.
+    # TODO: do proper line wrapping
+    enf p[1]' * (u_end + u_respan + 1 - (f_end(op_bits, extra) + f_respan(op_bits, extra)))
+        = p[1] * (v_join + v_split + v_loop + v_span + v_respan + 1 - (f_join(op_bits) + 
+        f_split(op_bits) + f_loop(op_bits) + f_span(op_bits) + 
+        f_respan(op_bits, extra)))
+
+
+# Enforces constraint for updating the block hash table.
+#
+# Constraint degree: 9
+ev block_hash_table_constraints([addr, op_bits[7], hasher[8], s0], [extra, p[4]]):
+    let op_bits_next = [op_bits[0]', op_bits[1]', op_bits[2]', op_bits[3]', op_bits[4]', op_bits[5]', op_bits[6]']
+    let extra_next = extra'
```

Comment:
actually, with the recent refactor of the IR, we should just be able to refer to `op_bits'` and `extra'`, so this shouldn't be necessary. I think this is used in a few other places too, so please search the file for "op_bits_next" and "extra_next" to update in the other places too